File: An II/Sem I/MAP/Seminar/Sem1/src/Model/QuickSort.java
package Model;

public class QuickSort extends AbstractSorter{

    @Override
    public void sort(int[] numbers) {
        quicksort(numbers, 0, numbers.length -1);
    }

    public void quicksort(int[] array, int begin, int end) {
        if(begin < end) {
            int partitionIndex = partition(array, begin, end);
            quicksort(array, begin, partitionIndex - 1);
            quicksort(array, partitionIndex + 1, end);
        }
    }

    private int partition(int[] array, int begin, int end) {
        int pivot = array[end];
        int i = begin - 1;
        for(int j = begin; j < end; j++) {
            if(array[j] <= pivot) {
                i++;
                int temp = array[i];
                array[i] = array[j];
                array[j] = temp;
            }
        }
        int temp = array[i + 1];
        array[i + 1] = array[end];
        array[end] = temp;

        return i + 1;
    }
}

File: An II/Sem I/MAP/Lab/Lab11/Lab11/Model/QuickSort.cs
namespace Lab11;

public class QuickSort : AbstractSorter
{
    private int partition(int[] elements, int left, int right)
    {
        var pivot = elements[right];
        var i = left - 1;
        for (var j = left; j < right; j++)
        {
            if (elements[j] >= pivot) continue;
            i++;
            (elements[i], elements[j]) = (elements[j], elements[i]);
        }
        (elements[i + 1], elements[right]) = (elements[right], elements[i + 1]);
        return i + 1;
    }
    
    private void quickSort(int[] elements, int left, int right)
    {
        if (left >= right) return;
        var pivot = partition(elements, left, right);
        quickSort(elements, left, pivot - 1);
        quickSort(elements, pivot + 1, right);
    }
    public override void sort(int[] array)
    {
        quickSort(array, 0, array.Length - 1);
    }
}